### PR TITLE
Audio: Crossover: Convert to module adapter

### PR DESCRIPTION
**Pull request by singalsu**
This patch contains basic changes to run crossover component as module adapter client in IPC3 system. Additional changes are needed for IPC4.

### CODE REVIEW

File: src/audio/crossover/crossover.c [hunk 1]
```diff
@@ -123,56 +124,61 @@ static int crossover_get_stream_index(struct sof_crossover_config *config,
...
 		sink = container_of(sink_list, struct comp_buffer, source_list);
 		sink_c = buffer_acquire(sink);
-
 		pipeline_id = sink_c->pipeline_id;
 		state = sink_c->sink->state;
 		buffer_release(sink_c);
-
-		if (state != dev->state)
+		if (state != dev->state) {
+			j++;
```
Comment (cujomalainey) [inline src/audio/crossover/crossover.c:150]: nice catch

File: src/audio/crossover/crossover.c [hunk 2]
```diff
@@ -123,56 +124,61 @@ static int crossover_get_stream_index(struct sof_crossover_config *config,
...
 		if (sinks[i]) {
 			comp_err(dev,
-				 "crossover_acquire_sinks(), multiple sinks from pipeline %d are assigned",
+				 "crossover_assign_sinks(), multiple sinks from pipeline %d are assigned",
 				 pipeline_id);
 			break;
 		}
 
 		sinks[i] = sink;
+		enabled[j++] = true;
```
Comment (cujomalainey) [inline src/audio/crossover/crossover.c:181]: these changes seem more like a bugfix rather than a IPC4 change, can you split into a different commit?

File: src/audio/crossover/crossover.c [hunk 3]
```diff
@@ -469,167 +461,85 @@ static int crossover_validate_config(struct comp_dev *dev,
...
+	struct comp_dev *dev = mod->dev;
+	struct audio_stream __sparse_cache *source = input_buffers[0].data;
 	uint32_t num_sinks;
 	uint32_t num_assigned_sinks = 0;
-	uint32_t frames = UINT_MAX;
-	uint32_t source_bytes, avail;
-	uint32_t sinks_bytes[SOF_CROSSOVER_MAX_STREAMS] = { 0 };
-
-	comp_dbg(dev, "crossover_copy()");
+	uint32_t frames = input_buffers[0].size;
```
Comment (cujomalainey) [inline src/audio/crossover/crossover.c:519]: this assumes all buffers have the same space, won't we overrun?

File: src/include/sof/audio/component.h [hunk 4]
```diff
@@ -715,6 +714,7 @@ void sys_comp_kpb_init(void);
 void sys_comp_multiband_drc_init(void);
 void sys_comp_selector_init(void);
 
+void sys_comp_module_crossover_interface_init(void);
```
Comment (btian1) [inline src/include/sof/audio/component.h:717]: seems this is not must, check up_down_mixer.c, it can be initialized with:
SOF_MODULE_INIT(up_down_mixer, sys_comp_module_up_down_mixer_interface_init);

File: src/audio/crossover/crossover.c [hunk 5]
```diff
@@ -642,215 +562,180 @@ static int crossover_copy(struct comp_dev *dev)
...
-		if (!sinks[i])
-			continue;
-		/*
-		 * WARNING: if a different thread happens to lock the same
-		 * buffers in different order, they can deadlock
-		 */
-		sinks_c[i] = buffer_acquire(sinks[i]);
-		avail = audio_stream_avail_frames(&source_c->stream,
-						  &sinks_c[i]->stream);
-		frames = MIN(frames, avail);
```
Comment (johnylin76) [inline src/audio/crossover/crossover.c:656]: This logic seems to be removed by the commit, is it intentional?
It looks important that determines the processed frames with regard to sink capacities in case of overrun.

File: src/audio/crossover/crossover.c [hunk 6]
```diff
@@ -642,215 +562,180 @@ static int crossover_copy(struct comp_dev *dev)
...
+		crossover_set_stream_params(sink, mod);
+	}
+#endif
 
 	/* Validate frame format and buffer size of sinks */
 	list_for_item(sink_list, &dev->bsink_list) {
 		sink = container_of(sink_list, struct comp_buffer, source_list);
 		sink_c = buffer_acquire(sink);
 
+		/* Note: Printing pipeline IDs for debug, these are with IPC4 zeros, probably due
```
Comment (johnylin76) [inline src/audio/crossover/crossover.c:646]: Does the bug you mentioned still exist under IPC4? If yes, is it worked-around by the commit, or is it non-critical?

File: src/include/sof/audio/component.h (hunk 4 above)
Comment (singalsu) [reply src/include/sof/audio/component.h:717]: Statically linked testbench needs this header, it's not using the advanced component introduce and initialize system.

File: src/audio/crossover/crossover.c (hunk 2 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:181]: I can't see the bug, is there one? The enabled vector is for correctly updating mod->output_buffers[i].size. I'm not sure why the multiple buffers are provided by module API like they are now with possibly different state than component. It would be nice if the slow list search could be avoided totally and only the buffers vector from module api could be used.

File: src/audio/crossover/crossover.c (hunk 3 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:529]: It should not happen. In this case case the function module_single_source_setup() checks the frames avail/free from all source and sink combinations: https://github.com/thesofproject/sof/blob/a38458dd2276fa24e37f5e60ef0a5613b03cbe00/src/audio/module_adapter/module_adapter.c#L778

File: src/audio/crossover/crossover.c (hunk 5 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:656]: Same as Curtis asked? The module adapter does this for clients. Though there's also the sinks[i] NULL case if states are not equal. I have not been able to create such test situation, could a NULL sink impact the avail/free. I'm not sure.

File: src/audio/crossover/crossover.c (hunk 6 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:646]: I'm now after vacation starting to study this again -- I hope I'll gain understanding of IPC4 to be able to address this.

File: src/audio/crossover/crossover.c (hunk 3 above)
Comment (cujomalainey) [reply src/audio/crossover/crossover.c:529]: Definitely not intuitive, mind leaving a comment explaining that?

File: src/audio/crossover/crossover.c [hunk 7]
```diff
@@ -106,12 +107,24 @@ static int crossover_get_stream_index(struct sof_crossover_config *config,
...
 	return -EINVAL;
 }
 
+static uint32_t crossover_get_sink_pipeline_id(struct comp_buffer __sparse_cache *sink_c)
+{
+#if CONFIG_IPC_MAJOR_4
+	// TODO: IPC4 renumbers pipeline_id:s due to pipelines count limitation, so can't use it.
+	// Use instead output pin index, cd->ipc4_cfg->output_pin.pin_index. The crossover
+	// topology blobs for IPC4 are now built with those indices.
+	return sink_c->sink->pipeline->pipeline_id;
```
Comment (singalsu) [inline src/audio/crossover/crossover.c:122]: This change is stll needed. The crossover configuration blobs for IPC4 need to be populated with pin indexes 0, 1, ... instead of pipeline_id:s.

**Comment by singalsu** [comment]
Crossover now runs with this PR correctly with IPC4 and IPC3, so changing this to non-draft. I'm able to run 2-way configuration with UPX board in nocodec mode. I play one stream and capture back from SSP loopback paths the band filtered filters versions those appear to be OK.

The other patches those belong to this work are:
- https://github.com/thesofproject/sof/pull/7883
- https://github.com/thesofproject/sof/pull/7884
- https://github.com/thesofproject/rimage/pull/170

File: src/audio/crossover/crossover.c (hunk 3 above)
Comment (kv2019i) [reply src/audio/crossover/crossover.c:628]: @singalsu can you add a comment?

File: src/include/sof/audio/component.h (hunk 4 above)
Comment (kv2019i) [reply src/include/sof/audio/component.h:719]: @btian1 I think the SOF_MODULE_INIT is in crossover.c -- looks correct o me.

**Review by kv2019i — COMMENTED** [review]
Thanks @singalsu ! I think there is one req-for-an-inline-comment outstanding, but otherwise looks good to go.

File: src/audio/crossover/crossover.c (hunk 3 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:632]: Thanks, yep, the comment is now added!

**Review by kv2019i — APPROVED** [review]

**Review by cujomalainey — APPROVED** [review]

File: src/audio/crossover/crossover.c [hunk 8]
```diff
@@ -642,215 +665,179 @@ static int crossover_copy(struct comp_dev *dev)
...
-		/*
-		 * WARNING: if a different thread happens to lock the same
-		 * buffers in different order, they can deadlock
-		 */
-		sinks_c[i] = buffer_acquire(sinks[i]);
-		avail = audio_stream_avail_frames(&source_c->stream,
-						  &sinks_c[i]->stream);
-		frames = MIN(frames, avail);
+	/* Process crossover */
+	if (frames) {
```
Comment (lyakh) [inline src/audio/crossover/crossover.c:669]: nitpick: an easier form would be
```
if (!frames)
    return -ENODATA;
```

File: src/audio/crossover/crossover.c [hunk 9]
```diff
@@ -642,215 +665,179 @@ static int crossover_copy(struct comp_dev *dev)
...
-								      dev->frames);
-			if (audio_stream_get_size(&sink_c->stream) < sink_period_bytes) {
-				comp_err(dev,
-					 "crossover_prepare(), sink %d buffer size %d is insufficient",
-					 sink_c->pipeline_id, audio_stream_get_size(&sink_c->stream));
-				ret = -ENOMEM;
-			}
 		}
 
+		audio_stream_init_alignment_constants(1, 1, &sink_c->stream);
```
Comment (lyakh) [inline src/audio/crossover/crossover.c:760]: this needs to be done even if the formats didn't match and we're going to error out?

File: src/audio/crossover/crossover.c (hunk 8 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:669]: Yep, done

File: src/audio/crossover/crossover.c (hunk 9 above)
Comment (singalsu) [reply src/audio/crossover/crossover.c:760]: Yep, ordered this way for more compact appear with my style but done now.

File: src/audio/crossover/crossover.c [hunk 10]
```diff
@@ -642,215 +666,179 @@ static int crossover_copy(struct comp_dev *dev)
...
-			}
 		}
 
 		buffer_release(sink_c);
-
 		if (ret < 0)
-			goto out;
+			return ret;
+
+		audio_stream_init_alignment_constants(1, 1, &sink_c->stream);
```
Comment (lyakh) [inline src/audio/crossover/crossover.c:764]: no, this cannot be done here - cannot use `sink_c` after releasing

**Review by btian1 — APPROVED** [review]

**Review by lyakh — APPROVED** [review]

**Comment by lgirdwood** [comment]
@wszypelt @lrudyX good to merge ? IIUC, crossover not tested by internal CI today.

**Comment by lrudyX** [comment]
> @wszypelt @lrudyX good to merge ? IIUC, crossover not tested by internal CI today.

It is waiting in queue. I have moved it to the top. We had some problems with DUTs.

**Comment by lrudyX** [comment]
@lgirdwood Internal CI result is pass :)
